File: c/easy/2154. Keep Multiplying Found Values by Two/main.c
# include <stdio.h>

# include <stdlib.h>

int partition (int * nums, int left, int right ){
    int i, j, pivo, temp;

    i = left -1;
    j = left;
    pivo = nums [ right ];

    for ( j; j < right; j++){
        if ( nums [ j ] <= pivo ){
            i++;
            temp = nums [ j ];
            nums [ j ] = nums [ i ];
            nums [ i ] = temp;
        }
    }

    temp = nums [ i + 1];
    nums [ i + 1 ] = nums [ right ];
    nums [ right ] = temp;

    return i + 1;
}

void quicksort ( int * nums, int left, int right ){
    int pi;
    
    if ( left < right ){
        pi = partition ( nums, left, right );
        quicksort ( nums, left, pi - 1);
        quicksort ( nums, pi + 1, right );
    }
}

int findFinalValue(int* nums, int numsSize, int original) {

    int i;

    if ( nums != NULL ){
        if ( numsSize > 0 ){
            quicksort ( nums, 0, numsSize - 1);

            for ( i = 0; i < numsSize; i++){
                if ( nums [ i ] == original ){
                    original *= 2;
                }
            }
        }
    }

    return original;   
    
}


 
void printArr ( int * nums, int numsSize ){
    int i = 0;
    if ( nums != NULL ){
        while ( i < numsSize ){
            printf ("%d ", nums [ i ]);
            i++;
        }
    }
    printf ("\n");
}


int main ( void ){
    int nums [ 5 ] = { 5, 3, 6, 1, 12};

    quicksort ( nums, 0, 4);

    printArr ( nums, 5);

    printf ("%d\n", findFinalValue( nums, 5, 3));

    return 0;

}

File: python/easy/2154. Keep Multiplying Found Values by Two/main.py
from typing import List

def quicksort ( arr: List [ int ], left: int, right: int ):
    if left < right:
        pi = partition ( arr, left, right )
        quicksort ( arr, left, pi -1 )
        quicksort ( arr, pi + 1, right )

def partition ( arr: List [ int ], left: int, right: int ) -> int:
    pivot = arr [ right ]

    i = left - 1

    for j in range ( left, right ):
        if arr [ j ] <= pivot:
            i += 1
            arr [ j ], arr [ i ] = arr [ i ], arr [ j ]
    
    arr [ i + 1], arr [ right ] = arr [ right ], arr [ i + 1]

    return i + 1

class Solution:

    # esse jeito foi para treinar quicksort

    def findFinalValue(self, nums: List[int], original: int) -> int:

        quicksort ( nums, 0, len ( nums ) - 1)

        for num in nums:
            if ( num == key ):
                original *= 2
        
        return original 

    def findFinalValue2 (self, nums: List[int], original: int) -> int:
        nums = set ( nums )

        while original in nums:
            original *= 2
        
        return original

sol = Solution ()

arr2 = [5, 3, 6, 1, 12]

print ( sol.findFinalValue ( arr2, 7))



        